### PR TITLE
Separate IPAM config/operational data in nw inspect 

### LINKED ISSUE #26799: Missing Gateway field in `bridge` network inspect
First time daemon is booted (no `docker0` bridge present, no or empty `/var/lib/docker/network/files/local-kv.db` file), IPAM.Config structure in network inspect output will not contain the Gateway field

```
$ docker network inspect bridge
[
    {
        "Name": "bridge",
        "Id": "2b5cc5f97b6de6a5e0533b1e2d5bec6284d900a9a82632ce782320a19db8b9e0",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.17.0.0/16"
                }
            ]
        },

```

Verified problem exists in `1.12.0` and in current master.

Given the `IPAM.Config` filed currently used to carry both configuration and operational data, the gateway address should always be there.

### PR DESCRIPTION
**Pull request by aboch**
Currently `IPAM.Config` entry in `docker network inspect` output  is being filled with the network ipam configuration data if present, otherwise with the ipam operationional data.

It was decided to finally separate config from operational state.
This means now the IPAM field in the network object will contain `Config` as well as `Data` field.

An excerpt from a `docker networ inspect` would look like this:

```
 "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "5.5.0.0/16",
                    "IPRange": "5.5.2.0/24",
                    "Gateway": "5.5.254.254",
                    "AuxiliaryAddresses": {
                        "my_router": "5.5.2.2",
                        "my_switch": "5.5.2.1"
                    }
                },
                {
                    "Subnet": "2001:db8:abcd:1234::/64",
                    "IPRange": "2001:db8:abcd:1234:5678::/80"
                }
            ],
            "Data": [
                {
                    "Subnet": "5.5.0.0/16",
                    "Gateway": "5.5.254.254"
                },
                {
                    "Subnet": "2001:db8:abcd:1234::/64",
                    "Gateway": "2001:db8:abcd:1234:5678::"
                }
            ]
        },
```

Fixes #26799 
Fixes #26933 

Signed-off-by: Alessandro Boch aboch@docker.com

### CODE REVIEW

**Comment by mrjana** [comment]
@aboch How do I know which one of those `IPAMConfig`s is actually config and which one is operation data? Isn't it better to separate Config and Operational data into two different slices and show them? Right now it will show up in as `IPAM.Config` which is not true.

**Comment by aboch** [comment]
@mrjana I initially thought of that, but from the issues about the network inspect it seems people are looking for the operational data in the current output, and this PR fixes that.
In fact currently for all the network created with no ipam config, we report the pure ipam data.

Given we fail the network creation if we cannot satisfy the config data, I am not sure whether there is a value in showing both. 
Consider the output may get pretty verbose if we list both config and operational, for then finding that both fields would have pretty much the same content most of the time.

I though of changing `Config` to `Data` to properly identify what we are returning, but that is an API change which I would refrain to do now. And would retain the status quo where user expects to find the ipam operational data in that field.

**Comment by mrjana** [comment]
@aboch Since we are talking about API representation it is fine if it is verbose as long as it provides unambiguous information. For example if somebody is writing migration code where they are reading the network state from one daemon to another, if they use the api in the current form then they can't recreate the network correctly because there is no way of knowing whether this is is user provided config or system generated operational data. I think we should follow the swarmkit api model of separating `Spec` from `Object`. User provided configuration is `Spec` which should never be modified by the system and the system generated state is in `Object`.

Since this is not going to be released in a patch release I think we can make the necessary API changes and support both versions based on API versioning design.

**Comment by aboch** [comment]
@mrjana
Yes, the migration is already broken, given today there is no way to know if what we read was a user config.

Ok then, let's reuse this PR to also take care of the config/operational separation.

**Comment by mrjana** [comment]
@aboch Thanks. Yes I was suggesting to make use of this PR to fix it properly since we are changing this part of the API behavior.

**Comment by aboch** [comment]
@mrjana Took care of your comment, added the new `IPAM.Data` field and now `IPAM.Config` really represents the user configuration only.

PTAL when you get a chance.

**Comment by aboch** [comment]
experimental to pass requires docker-py changes

**Comment by aboch** [comment]
Finally all green. @mrjana PTAL.

**Comment by mrjana** [comment]
LGTM

File: api/types/network/network.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ type IPAM struct {
 	Driver  string
 	Options map[string]string //Per network IPAM driver options
 	Config  []IPAMConfig
+	Data    []IPAMConfig // IPAM operational data
```
Comment (cpuguy83) [inline api/types/network/network.go]: `Data` doesn't seem like a great name... but I'm blanking on a good one atm.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

**Comment by thaJeztah** [comment]
@aboch would it be an idea to use the same concept as the service API; i.e. have a Spec field at the top-level, and anything inside that is the "provided" configuration

**Comment by cpuguy83** [comment]
I think this change needs to be done in the swagger spec now?

File: api/types/network/network.go (hunk 1 above)
Comment (aboch) [inline api/types/network/network.go]: `Operational`, `OperationalData`, `OperationalState` ?
I am not good at names either. We need to somehow convey this info is the ipam operational state.
Hopefully somebody can propose a good one.

**Comment by aboch** [comment]
@thaJeztah 
This change is backward compatible for the tools which already check the `Config` field.
I am not sure if we want to change the network inspect o/p structure now.

**Comment by aboch** [comment]
@cpuguy83 Thanls. I will take care of the swagger specs

**Comment by LK4D4** [comment]
ping @aboch 
Is this still a thing or should we close it?
Thanks!

**Comment by aboch** [comment]
It's still a thing, but do not have time to invastgate/fix the docker-py test failures now.

**Comment by icecrime** [comment]
Let's reopen when the PR is complete then. Thanks.

**Pull request closed without merging**